### PR TITLE
RAC-232: Add custom react component in job tracker grid

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**
How Progress is calculated ?
- All steps that is trackable need to implement TrackableStep, this interface expose `isTrackable` function
- All TrackableStep is responsible to set functions `setTotalItems` and `incrementProcessedCount`

Actually we have 4 type of Steps:
- ItemStep => is trackable only if reader is instance of TrackableItemReaderInterface
- ValidatorStep => It only validate imported file => For me this complicated to have a real time estimation. Actually it's only one item to process. This step is really quick it should not be problematic for users.
- TaskletStep => is trackable only if tasklet is instance of TaskletTrackableInterface. Tasklet are trackable when it can count total of item, in case of use Reader, tasklet is trackable when reader is instance of TrackableItemReaderInterface
- MassEditStep => badly named => it only clear temporary files (MassEdit used ItemStep). Like ValidatorStep it's only one item to process.

Potential problems :
- In front we should take care about status before calculating percentage (in case of developer miss to increment counter in some specific case)
- Step execution is not always save into database at each loop in Tasklet => we should change it

Questions : 
Actually, JobExecutionController is called by view each X second, should we expose the data in this controller or should we keep exposing data in JobExecutionProgressController ?

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Tool/Component/Batch/Model/StepExecution.php [hunk 1]
```diff
@@ -77,6 +77,11 @@ class StepExecution
     /** @var array */
     private $summary = [];
 
+    private $trackingData = [
+        'totalItems' => 0,
+        'processCount' => 0
```
Comment (juliensnz) [inline src/Akeneo/Tool/Component/Batch/Model/StepExecution.php:82]: What do you think about `processedItems` instead?

File: src/Akeneo/Tool/Component/Batch/Model/StepExecution.php [hunk 2]
```diff
@@ -220,6 +225,39 @@ public function setWriteCount($writeCount)
...
+
+    public function incrementProcessedCount($increment = 1)
+    {
+        $this->trackingData['processCount'] += $increment;
+        if ($this->trackingData['processCount'] > $this->getTotalItems()) {
+            $this->setTotalItems($this->trackingData['processCount']);
+        }
+    }
+
+    public function getProcessCount(): int
```
Comment (juliensnz) [inline src/Akeneo/Tool/Component/Batch/Model/StepExecution.php:246]: same here

File: src/Akeneo/Tool/Component/Batch/Step/ItemStep.php [hunk 3]
```diff
@@ -104,6 +110,10 @@ public function doExecute(StepExecution $stepExecution)
 
         $this->initializeStepElements($stepExecution);
 
+        if ($this->reader instanceof TrackableItemReaderInterface) {
```
Comment (juliensnz) [inline src/Akeneo/Tool/Component/Batch/Step/ItemStep.php:113]:
```suggestion
        if ($this->isTrackable()) {
```

File: src/Akeneo/Tool/Component/Connector/Step/ValidatorStep.php [hunk 4]
```diff
@@ -41,8 +42,10 @@ public function __construct(
      */
     protected function doExecute(StepExecution $stepExecution)
     {
+        $stepExecution->setTotalItems(1);
```
Comment (juliensnz) [inline src/Akeneo/Tool/Component/Connector/Step/ValidatorStep.php:45]: Are we sure we want to do this? Maybe we should not make it trackable and handle this case in the frontend?

**Review by juliensnz — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Component/Product/Job/ComputeFamilyVariantStructureChangesTasklet.php [hunk 5]
```diff
@@ -90,6 +91,11 @@ public function setStepExecution(StepExecution $stepExecution)
         $this->stepExecution = $stepExecution;
     }
 
+    public function isTrackable(): bool
```
[the remark this answers is not in the document]
Comment (StevenVAIDIE) [reply src/Akeneo/Pim/Enrichment/Component/Product/Job/ComputeFamilyVariantStructureChangesTasklet.php:94]: Not really, the tasklet can implement interface TaskletTrackableInterface with a reader not trackable

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Controller/InternalApi/JobInstanceController.php [hunk 6]
```diff
@@ -521,9 +529,24 @@ public function getJobNamesAction(Request $request)
...
+            foreach ($this->jobRegistry->allByTypeGroupByConnector($jobType) as $connector => $jobs) {
+                foreach ($jobs as $key => $job) {
+                    /** @var Job $job */
+                    if (in_array($job->getName(), $jobNotVisible)) {
+                        continue;
+                    }
+
+                    foreach ($job->getStepNames() as $stepName) {
+                        $step = $job->getStep($stepName);
+                        if(! $step->isTrackable()) {
```
[the remark this answers is not in the document]
Comment (StevenVAIDIE) [reply src/Akeneo/Platform/Bundle/ImportExportBundle/Controller/InternalApi/JobInstanceController.php:545]: I totally forget to clean this file, this code help me to show step not trackable

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/public/js/job-status-badge.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import React, {useEffect, useState} from "react";
+
+const JobStatusBadge = ({jobId, status}: { jobId: string; status: string }) => {
+  const [jobStatus, setStatus] = useState({
+    status: status,
+    currentStep: 0,
+    totalSteps: 0
+  });
+
+  useEffect(() => {
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/public/js/job-status-badge.tsx:11]: We should only fetch data when status is in progress

**Comment by StevenVAIDIE** [comment]
Real code merged here: https://github.com/akeneo/pim-community-dev/pull/13121

**Pull request closed without merging**